storage/tsdb/tsdb.go

```diff
@@ -15,322 +15,208 @@ package tsdb
+func (s *ReadyStorage) watchConfigForChanges(configPath string) error {
+	s.logger.Log("msg", "Starting to watch configuration for changes", "configPath", configPath)
 
-	// Initialize V3IO Adapter
-	cfgpath := os.Getenv("V3IO_FILE_PATH")
-	if cfgpath == "" {
-		cfgpath = "/etc/v3io/v3io-tsdb.yaml"
-	}
-	cfg, err := config.LoadConfig(cfgpath)
+	// get initial file mtime
+	lastConfigFileInfo, err := os.Stat(configPath)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	if !cfg.Disabled {
-		v3ioadp, err = promtsdb.NewV3ioProm(cfg, nil, nil) //NEW
-		if err != nil {
-			return nil, err
-		}
-	}
+	// in the background, check for mtime changes and reload config
+	go func() error {
 
-	// TODO: remove the tsdb.DB init
+		for !s.closed {
 
-	if opts.MinBlockDuration > opts.MaxBlockDuration {
-		opts.MaxBlockDuration = opts.MinBlockDuration
-	}
-	// Start with smallest block duration and create exponential buckets until the exceed the
-	// configured maximum block duration.
-	rngs := tsdb.ExponentialBlockRanges(int64(time.Duration(opts.MinBlockDuration).Seconds()*1000), 10, 3)
-
-	for i, v := range rngs {
-		if v > int64(time.Duration(opts.MaxBlockDuration).Seconds()*1000) {
-			rngs = rngs[:i]
-			break
-		}
-	}
+			// wait a bit
+			time.Sleep(3 * time.Second)
 
-	db, err := tsdb.Open(path, l, r, &tsdb.Options{
-		WALFlushInterval:  10 * time.Second,
-		RetentionDuration: uint64(time.Duration(opts.Retention).Seconds() * 1000),
-		BlockRanges:       rngs,
-		NoLockfile:        opts.NoLockfile,
-	})
-	if err != nil {
-		return nil, err
-	}
-	return db, nil
-}
+			// get current file mtime
+			currentConfigFileInfo, err := os.Stat(configPath)
+			if err != nil {
+				level.Warn(s.logger).Log("msg", "Failed to get config file info", "err", err.Error())
+				continue
+			}
 
-// StartTime implements the Storage interface.
-func (a adapter) StartTime() (int64, error) {
-	var startTime int64
+			// compare mtimes
+			if lastConfigFileInfo.ModTime() != currentConfigFileInfo.ModTime() {
+				lastConfigFileInfo = currentConfigFileInfo
 
-	if len(a.db.Blocks()) > 0 {
-		startTime = a.db.Blocks()[0].Meta().MinTime
-	} else {
-		startTime = int64(time.Now().Unix() * 1000)
-	}
+				s.logger.Log("msg", "Config file modification detected, recreating adapter", "configPath", configPath)
 
-	// Add a safety margin as it may take a few minutes for everything to spin up.
-	return startTime + a.startTimeMargin, nil
-}
+				newV3ioPromAdapter, err := s.createV3ioPromAdapater(configPath)
+				if err != nil {
+					level.Warn(s.logger).Log("msg", "Failed to create new v3io adapter", "err", err.Error())
+					continue
+				}
 
-func (a adapter) Querier(_ context.Context, mint, maxt int64) (storage.Querier, error) {
-	q, err := a.db.Querier(mint, maxt)
-	if err != nil {
-		return nil, err
-	}
-	return querier{q: q}, nil
-}
+				// swap old pointer with new (atomic)
+				s.v3ioPromAdapter = newV3ioPromAdapter
```

Comment:
Shouldn't this be guarded by a lock? Since it's running on background the old adapter instance could be in use.

Reply:
I thought about this. Swapping the pointer itself is atomic - there's no chance that a reader will read a "partially" updated pointer.

Now lets assume that a query / append happens at exactly the same time as an update. It will either happen before the pointer update and use the old object (I don't destroy it, I leave that to GC and GC won't destroy it until the query/append is done). Or after the pointer update - in which case it'll use the new object and the old should be marked for GC.

At least in theory.

Reply:
I'd be more worried about two appenders working concurrently on the same data. This could happen either (1) as an edge case similar to what @pavius described or (2) because Prometheus keeps the original instance for more than just the life of an operation (e.g. I see that it stores it in a "scrape loop", which sounds like a long-living thing).

Reply:
We need this functionality for 1.9.* to work with TSDB - we can assume for now that this functionality should not be used if Prometheus is scraping.